File: 10-regular-expression-matching/10-regular-expression-matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        return solve(s, p);
    }
private:
    // Rahul Sir
    bool solve(string s, string p) {
        if(p == ".*")
            return true;
        int n = s.size(), m = p.size();
        vector<vector<bool>> dp(m + 1, vector<bool>(n + 1, false));
        
        dp[0][0] = true;
        for(int i = 1; i <= m; ++i) {
            if(p[i - 1] == '*')
                dp[i][0] = dp[i - 2][0];
        }
        
        for(int i = 1; i <= m; ++i) {
            for(int j = 1; j <= n; ++j) {
                
                if((p[i - 1] == '.') || (p[i - 1] == s[j - 1]))
                    dp[i][j] = dp[i - 1][j - 1];
                
                else if(p[i - 1] == '*') {
                    dp[i][j] = dp[i - 2][j];
                    
                    if((p[i - 2] == s[j - 1]) || (p[i - 2] == '.'))
                        dp[i][j] = dp[i - 2][j] || dp[i][j - 1];
                }
            }
        }
        return dp[m][n];
    }
}; 

File: 10-regular-expression-matching/10-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if p == ".*":
            return True
        
        m = len(p)
        n = len(s)
        
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        
        dp[0][0] = True
        
        for i in range(1, m + 1):
            if p[i - 1] == '*':
                dp[i][0] = dp[i - 2][0]
                
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if p[i - 1] == '.' or p[i - 1] == s[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                
                elif p[i - 1] == '*':
                    dp[i][j] = dp[i - 2][j]
                    
                    if p[i - 2] == '.' or p[i - 2] == s[j - 1]:
                        dp[i][j] = dp[i][j] or dp[i][j - 1]
                        
        return dp[m][n]